fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -0,0 +1,356 @@
+        // value(s) are.
+        builder.setCount(partition.getCount() - 1);
+
+        if (builder.getCount() < 0) {
+            // should never happen
+            throw new RecordCoreException("Issue updating Lucene partition metadata (resulting count < 0)", "partitionId", partition.getId());
+        }
+        savePartitionMetadataAndUpdateCache(grouping, builder);
+    }
+
+    /**
+     * given a grouping key and a partition metadata object builder, build and save the partition metadata object in the db, and update the local cache.
+     *
+     * @param grouping grouping key
+     * @param builder proto builder
+     */
+    private void savePartitionMetadataAndUpdateCache(final Tuple grouping, final LucenePartitionInfoProto.LucenePartitionInfo.Builder builder) {
+        byte[] partitionKey = state.indexSubspace.subspace(grouping).pack(Tuple.from(PARTITION_META_SUBSPACE, builder.getId()));
+        LucenePartitionInfoProto.LucenePartitionInfo updatedPartition = builder.build();
+        state.context.ensureActive().set(partitionKey, updatedPartition.toByteArray());
+
+        // update cache
+        List<LucenePartitionInfoProto.LucenePartitionInfo> partitions = partitionMetaInfoCache.get(grouping);
+
+        // partition ids are in order, in the list
+        for (int i = 0; i < partitions.size(); i++) {
+            if (partitions.get(i).getId() == updatedPartition.getId()) {
+                partitions.set(i, updatedPartition);
+                return;
+            }
+        }
+
+        // should be unreachable
+    }
+
+    /**
+     * Selects the correct partition to assign a document to.
+     *
+     * @param partitions candidate partitions
+     * @param timestamp timestamp of the document
+     * @return assigned partition
+     */
+    private LucenePartitionInfoProto.LucenePartitionInfo selectWritePartition(final List<LucenePartitionInfoProto.LucenePartitionInfo> partitions, final long timestamp) {
+
+        // candidate partitions must include at least one
+        Preconditions.checkNotNull(partitions);
+        Preconditions.checkArgument(!partitions.isEmpty());
+
+        // ----------------------------------------------------------
+        // The partitioning rules are as follows (per requirements doc):
+        //
+        // When adding an email
+        //
+        // If its date is within a range of an existing partition it will get added to that partition.
+        // If it is between two partitions, it will get added to the smaller one, expanding that partitions date range
+        // If it is older than the oldest partition’s min date it will get added there.
+        // If it is newer than the newest partition’s max date it will get added there.
+        //
+        // If the partition that we just added to is larger than “max” size we will probabilistically split it, with a growing probability as the size increases.
```

Comment:
theoretical in the context of mail, but this can happen in other contexts, or in the case of an improperly built test population tool etc. Not sure how to handle this, will think about it.